py_bentoctl/auth_helper.py

```diff
@@ -96,12 +106,119 @@ def fetch_existing_client_id(token: str, client_id: str) -> Optional[str]:
 
     for client in existing_clients:
         if client["clientId"] == client_id:
-            warn(f"    Found existing client: {client_id}; using that.")
+            if verbose:
+                warn(f"    Found existing client: {client_id}; using that.")
             return client["id"]
 
     return None
 
 
+def fetch_existing_client_role(token: str, client_id: str, role_name: str, verbose: bool = True) -> Optional[dict]:
+    client_roles_endpoint = f"{KC_ADMIN_API_CLIENTS_ENDPOINT}/{client_id}/roles"
+
+    existing_role_res = keycloak_req(f"{client_roles_endpoint}/{role_name}", bearer_token=token)
+    if not existing_role_res.ok:
+        return
+
+    if verbose:
+        warn(f"    Found existing role: {role_name}; using that.")
+    return existing_role_res.json()
+
+
+def fetch_existing_group_rep_or_exit(
+        token: str,
+        group_name: str,
+        parent_rep: dict | None = None,
+        verbose: bool = True
+        ) -> Optional[dict]:
+    endpoint = KC_ADMIN_API_GROUP_ENDPOINT
+    if parent_rep:
+        endpoint = f"{KC_ADMIN_API_GROUP_ENDPOINT}/{parent_rep['id']}/children"
+    existing_groups_res = keycloak_req(endpoint, bearer_token=token)
+
+    if not existing_groups_res.ok:
+        err(f"    Failed to fetch group id associated with name: {group_name}")
+        exit(1)
+
+    existing_groups = existing_groups_res.json()
+    for group in existing_groups:
+        if group["name"] == group_name:
+            if verbose:
+                warn(f"    Found existing group: {group['path']} ; using that.")
+            return group
+
+    return None
+
+
+def create_client_role_or_exit(token: str, client_id: str, role_name: str) -> Optional[dict]:
+    # Check if client role alread exists
+    if existing_role := fetch_existing_client_role(token, client_id, role_name):
+        return existing_role
+
+    # Create client role if needed
+    client_roles_endpoint = f"{KC_ADMIN_API_CLIENTS_ENDPOINT}/{client_id}/roles"
+    res = keycloak_req(client_roles_endpoint, bearer_token=token, method="post", json={
+        "clientRole": True,
+        "name": role_name,
+    })
+
+    if not res.ok:
+        err(f"    Failed to create {client_id} client role : {role_name}; {res.status_code}")
+        exit(1)
+
+    # role creation response returns no data, fetch the created RoleRepresentation for later use
+    created_role_rep = fetch_existing_client_role(token, client_id, role_name, verbose=False)
+    warn(f"    Created client role: {role_name}.")
```

Comment:
is this a warning? other places use a created + done approach

Reply:
done